File: М10_ServerVMS/vmsserver-go/psimplatform/eventindex.go
package psimplatform

// eventindex — the event "database", which is a cache. A platform job.
//
// Events are observations: written by the worker that holds a unit's epoch,
// into that unit's bucket on its server's resource. Cross-unit search needs
// an index over all of them, and this is it: one per cluster (or per box),
// holding a table it can rebuild entirely by re-reading every resource's
// buckets. (The Python version keeps the table in SQLite; this one keeps it
// in memory — it is a cache either way, and the standard library has no
// SQLite.) It knows which subsystems exist by what it finds in the
// resources' heartbeats. It knows nothing about what an event means: `cam`
// is a field an event may carry, indexed if present.
//
//	Rebuild(resources)   read every subsystem's closed buckets on every resource and index them
//	Tail(resources)      the same, for buckets it has not seen (by (server, path))
//	Query(...)           subsystem, unit, cam, kind, time window; Unreachable names silent resources

import (
	"fmt"
	"io"
	"net/http"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// ResourceReader: HTTP against the resource job; tests substitute a directory reader.
type ResourceReader interface {
	Buckets(url, sub, unit string) ([]Bucket, error)
	Events(url string, b Bucket) ([]Event, error)
	Mirrored(url, server string) ([]Bucket, error)
	MirroredEvents(url, server string, b Bucket) ([]Event, error)
}

type HTTPResourceReader struct{ Client *http.Client }

func NewHTTPResourceReader() *HTTPResourceReader {
	return &HTTPResourceReader{&http.Client{Timeout: 3 * time.Second}}
}

func (r *HTTPResourceReader) get(u string) ([]byte, error) {
	resp, err := r.Client.Get(u)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("GET %s: %d", u, resp.StatusCode)
	}
	return io.ReadAll(resp.Body)
}

func parseEvents(raw []byte) []Event {
	var out []Event
	for _, l := range strings.Split(string(raw), "\n") {
		if strings.TrimSpace(l) == "" {
			continue
		}
		var e Event
		if err := jsonUnmarshal([]byte(l), &e); err == nil {
			out = append(out, e)
		}
	}
	return out
}

func (r *HTTPResourceReader) Buckets(url, sub, unit string) ([]Bucket, error) {
	raw, err := r.get(url + "/buckets/" + sub + "/" + unit)
	if err != nil {
		return nil, err
	}
	return ParseBucketLines(strings.NewReader(string(raw)))
}

func (r *HTTPResourceReader) Events(url string, b Bucket) ([]Event, error) {
	raw, err := r.get(url + "/events/" + b.Path)
	if err != nil {
		return nil, err
	}
	return parseEvents(raw), nil
}

func (r *HTTPResourceReader) Mirrored(url, server string) ([]Bucket, error) {
	raw, err := r.get(url + "/mirrored/" + server)
	if err != nil {
		return nil, err
	}
	return ParseBucketLines(strings.NewReader(string(raw)))
}

func (r *HTTPResourceReader) MirroredEvents(url, server string, b Bucket) ([]Event, error) {
	raw, err := r.get(url + "/events/" + MirrorDir + "/" + server + "/" + b.Path)
	if err != nil {
		return nil, err
	}
	return parseEvents(raw), nil
}

// IndexedEvent is one row of the index. Cam is nil when the event named no
// cam and its unit is not numeric.
type IndexedEvent struct {
	Subsystem, Unit string
	Cam             *int
	Epoch           int
	T               float64
	Kind, Server    string
	Bucket          string
	Fenced          bool
	Fields          map[string]any
}

// CamIs is the test's shorthand: a Cam that is set and equals n.
func (e IndexedEvent) CamIs(n int) bool { return e.Cam != nil && *e.Cam == n }

type IndexReport struct {
	Added       int
	Unreachable []string
	FromMirror  []string
	Segments    int
}

type QueryResult struct {
	Events []IndexedEvent
	State  string
}

// Query parameters; nil pointers mean "any".
type Query struct {
	T0, T1        float64
	Cam           *int
	Kind          *string
	Subsystem     *string
	Unit          *string
	CurrentEpochs map[[2]string]int // (subsystem, unit) -> epoch: fencing is per unit
	Limit         int
}

func IntPtr(n int) *int       { return &n }
func StrPtr(s string) *string { return &s }

type EventIndex struct {
	mu              sync.Mutex
	Reader          ResourceReader
	Wall            Clock
	LostAfter       float64
	State           string
	IndexedSegments int
	seen            map[[2]string]bool
	rows            []IndexedEvent
}

func NewEventIndex(reader ResourceReader, wall Clock) *EventIndex {
	if wall == nil {
		wall = Wall()
	}
	return &EventIndex{Reader: reader, Wall: wall, LostAfter: 45, State: "empty", seen: map[[2]string]bool{}}
}

// Rebuild from nothing: what a failed-over instance does first.
func (x *EventIndex) Rebuild(resources map[string]ResourceHeartbeat) IndexReport {
	x.mu.Lock()
	x.seen, x.rows, x.IndexedSegments = map[[2]string]bool{}, nil, 0
	x.mu.Unlock()
	return x.tail(resources, true)
}

func (x *EventIndex) Tail(resources map[string]ResourceHeartbeat) IndexReport {
	return x.tail(resources, false)
}

func camOf(e Event, unit string) *int {
	if v, ok := e["cam"]; ok {
		n := int(ToFloat(v))
		return &n
	}
	if n, err := strconv.Atoi(unit); err == nil {
		return &n
	}
	return nil
}

func (x *EventIndex) insert(server string, b Bucket, events []Event) int {
	x.mu.Lock()
	defer x.mu.Unlock()
	n := 0
	for _, e := range events {
		fields := map[string]any{}
		for k, v := range e {
			if k != "t" && k != "kind" && k != "cam" {
				fields[k] = v
			}
		}
		x.rows = append(x.rows, IndexedEvent{b.Subsystem, b.Unit, camOf(e, b.Unit), b.Epoch, e.T(), e.Kind(), server, b.Path, false, fields})
		n++
	}
	x.seen[[2]string{server, b.Path}] = true
	x.IndexedSegments++
	return n
}

func (x *EventIndex) hasSeen(server, path string) bool {
	x.mu.Lock()
	defer x.mu.Unlock()
	return x.seen[[2]string{server, path}]
}

func (x *EventIndex) tail(resources map[string]ResourceHeartbeat, rebuild bool) IndexReport {
	if rebuild {
		x.State = "catching up"
	}
	now := x.Wall()
	rep := IndexReport{Unreachable: []string{}, FromMirror: []string{}}
	live := map[string]ResourceHeartbeat{}
	var servers []string
	for s, hb := range resources {
		servers = append(servers, s)
		if now-hb.Ts <= x.LostAfter {
			live[s] = hb
		}
	}
	sort.Strings(servers)
	for _, server := range servers {
		hb := resources[server]
		if _, ok := live[server]; !ok {
			if added, ok := x.fromMirror(server, live); ok {
				rep.FromMirror = append(rep.FromMirror, server)
				rep.Added += added
			} else {
				rep.Unreachable = append(rep.Unreachable, server)
			}
			continue
		}
		if err := x.indexServer(server, hb, &rep); err != nil { // fresh heartbeat, server not answering
			rep.Unreachable = append(rep.Unreachable, server)
		}
	}
	state := "live"
	if len(rep.Unreachable) > 0 {
		state += "; " + strings.Join(rep.Unreachable, ", ") + " unreachable"
	}
	if len(rep.FromMirror) > 0 {
		state += "; " + strings.Join(rep.FromMirror, ", ") + " from mirror"
	}
	x.State = state
	rep.Segments = x.IndexedSegments
	return rep
}

func (x *EventIndex) indexServer(server string, hb ResourceHeartbeat, rep *IndexReport) error {
	var subs []string
	for s := range hb.Units {
		subs = append(subs, s)
	}
	sort.Strings(subs)
	for _, sub := range subs {
		for _, unit := range hb.Units[sub] {
			bs, err := x.Reader.Buckets(hb.URL, sub, unit)
			if err != nil {
				return err
			}
			for _, b := range bs {
				if b.Events == 0 || x.hasSeen(server, b.Path) {
					continue
				}
				evs, err := x.Reader.Events(hb.URL, b)
				if err != nil {
					return err
				}
				rep.Added += x.insert(server, b, evs)
			}
		}
	}
	return nil
}

// fromMirror: a silent server's closed buckets, from whichever live peer
// holds copies. Rows are inserted under the REAL server: only the source
// differed. Never used while the resource answers.
func (x *EventIndex) fromMirror(server string, live map[string]ResourceHeartbeat) (int, bool) {
	var holders []ResourceHeartbeat
	for _, hb := range live {
		if _, ok := hb.Mirrors[server]; ok {
			holders = append(holders, hb)
		}
	}
	if len(holders) == 0 {
		return 0, false
	}
	added := 0
	for _, hb := range holders {
		bs, err := x.Reader.Mirrored(hb.URL, server)
		if err != nil {
			continue // that peer is not answering either
		}
		for _, b := range bs {
			if b.Events == 0 || x.hasSeen(server, b.Path) {
				continue
			}
			evs, err := x.Reader.MirroredEvents(hb.URL, server, b)
			if err != nil {
				break
			}
			added += x.insert(server, b, evs)
		}
	}
	return added, true
}

func (x *EventIndex) Query(q Query) QueryResult {
	x.mu.Lock()
	defer x.mu.Unlock()
	limit := q.Limit
	if limit == 0 {
		limit = 1000
	}
	var out []IndexedEvent
	for _, r := range x.rows {
		if r.T < q.T0 || r.T >= q.T1 {
			continue
		}
		if q.Cam != nil && !r.CamIs(*q.Cam) {
			continue
		}
		if q.Kind != nil && r.Kind != *q.Kind {
			continue
		}
		if q.Subsystem != nil && r.Subsystem != *q.Subsystem {
			continue
		}
		if q.Unit != nil && r.Unit != *q.Unit {
			continue
		}
		if cur, ok := q.CurrentEpochs[[2]string{r.Subsystem, r.Unit}]; ok {
			r.Fenced = r.Epoch < cur
		}
		out = append(out, r)
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].T < out[j].T })
	if len(out) > limit {
		out = out[:limit]
	}
	if out == nil {
		out = []IndexedEvent{}
	}
	return QueryResult{out, x.State}
}

// Forget: retention on a resource removed a segment; its events go with it.
func (x *EventIndex) Forget(server string, paths []string) int {
	x.mu.Lock()
	defer x.mu.Unlock()
	gone := map[string]bool{}
	for _, p := range paths {
		gone[p] = true
		delete(x.seen, [2]string{server, p})
	}
	var keep []IndexedEvent
	for _, r := range x.rows {
		if !(r.Server == server && gone[r.Bucket]) {
			keep = append(keep, r)
		}
	}
	x.rows = keep
	return len(paths)
}

// ToMap renders an indexed event as the console's JSON row.
func (e IndexedEvent) ToMap() map[string]any {
	m := map[string]any{"subsystem": e.Subsystem, "unit": e.Unit, "epoch": e.Epoch, "t": e.T, "kind": e.Kind,
		"server": e.Server, "bucket": e.Bucket, "fenced": e.Fenced}
	if e.Cam != nil {
		m["cam"] = *e.Cam
	} else {
		m["cam"] = nil
	}
	for k, v := range e.Fields {
		m[k] = v
	}
	return m
}

File: М10_ServerVMS/vmsserver/vmsplatform/eventindex.py
"""eventindex — the event "database", which is a cache. A platform job.

Events are observations: written by the worker that holds a unit's epoch,
into that unit's bucket on its server's resource (vmsplatform.events).
Cross-unit search needs an index over all of them, and this is it: one
per cluster (or per box), holding a SQLite table it can rebuild entirely
by re-reading every resource's buckets. It knows which subsystems exist by
what it finds in the resources' heartbeats; a new one is indexed the pass
after it starts writing, with no change here. It knows nothing about what
an event means: `cam` is a field an event may carry, indexed if present.

Its two properties are the controller's, in the form that matters here:
it holds nothing it cannot rebuild, and nothing running depends on it.
No controller writes events. A failed-over eventindex says *catching up*
until its rebuild is done rather than answering short.

    rebuild(resources)   read every subsystem's closed buckets on every resource and index them
    tail(resources)      the same, for buckets it has not seen (by (server, path))
    query(...)           subsystem, unit, cam (a field an event may carry), kind, time window; `unreachable` names silent resources
"""
# ================================================================================================
# NOTES — what every part of this file does and why (kept beside the code, not in a separate document)
# ================================================================================================
# # eventindex.py — the event "database", which is a cache: a SQLite index over every subsystem's buckets on
# every resource
#
# **Role in the module.** Lesson 3, the cluster-side reader of `events.py`. Events are written by workers
# into per-unit buckets on their server's resource; cross-unit search needs one index over all of them, and
# this is it: one per cluster (or per box), a SQLite table it can rebuild entirely by re-reading every
# resource's closed buckets over the resource's HTTP routes. It discovers subsystems and units from the
# resource heartbeats' `units` field, so a new subsystem is indexed the pass after it starts writing, with
# no change here. It knows nothing about what an event means: `cam` is a field an event may carry, indexed
# if present. Its two properties are the controller's in the form that matters here: it holds nothing it
# cannot rebuild, and nothing running depends on it. No controller writes events. A failed-over instance
# says *catching up* until its rebuild is done rather than answering short. `console.SpecConsole` serves
# `/events` from an instance when one is passed as `index`; none of the 47 tests construct one (it is
# exercised on the box / in М11).
#
# ## Module-level names
# None.
#
# ### `query(self, t0, t1, cam=None, kind=None, subsystem=None, unit=None, current_epochs=None, limit=1000)
# ->
# dict` Time window plus optional equality filters, ordered by `t`, limited. `current_epochs` is
# `{(subsystem, unit): epoch}` — fencing is per unit, and only the unit's own subsystem knows its current
# epoch; the index just compares and sets `fenced: epoch < current`. Each result row is `{subsystem, unit,
# cam, epoch, t, kind, server, bucket, fenced, **fields}`; the reply is `{events, state}` so a caller can
# tell a short answer during catch-up from a complete one. The console builds `current_epochs` from
# `<sub>/epoch/*`.
#
# ## Notes
# - The `cam` column exists so a VMS timeline can ask "everything about camera 7 from any subsystem" without
#   the index knowing what a camera is; the console's `/events?cam=` maps onto it.
# - Nothing is ever updated in place: rows are inserted per bucket and deleted per bucket, matching the
#   resource's file-level retention.
# ================================================================================================
from __future__ import annotations

import json
import sqlite3
import time
import urllib.request

from .events import Bucket
from .resource import bucket_from_line


# HTTP against the resource job (`resource.serve`); tests would substitute a directory reader with the same
# four methods.
class ResourceReader:
    """HTTP against the resource job; tests substitute a directory reader."""
    def __init__(self, timeout: float = 3.0): self.timeout = timeout

    # `GET <url>/buckets/<sub>/<unit>`, parsed with `bucket_from_line`.
    def buckets(self, url: str, sub: str, unit: str) -> list[Bucket]:
        with urllib.request.urlopen(f"{url}/buckets/{sub}/{unit}", timeout=self.timeout) as r:
            return [bucket_from_line(l) for l in r.read().decode().splitlines() if l.strip()]

    # `GET <url>/events/<b.path>`, one dict per line.
    def events(self, url: str, b: Bucket) -> list[dict]:
        with urllib.request.urlopen(f"{url}/events/{b.path}", timeout=self.timeout) as r:
            return [json.loads(l) for l in r.read().decode().splitlines() if l.strip()]

    # `GET <url>/mirrored/<server>` — a peer's copies of `server`'s buckets.
    def mirrored(self, url: str, server: str) -> list[Bucket]:
        with urllib.request.urlopen(f"{url}/mirrored/{server}", timeout=self.timeout) as r:
            return [bucket_from_line(l) for l in r.read().decode().splitlines() if l.strip()]

    # `GET <url>/events/.mirror/<server>/<b.path>`.
    def mirrored_events(self, url: str, server: str, b: Bucket) -> list[dict]:
        with urllib.request.urlopen(f"{url}/events/.mirror/{server}/{b.path}", timeout=self.timeout) as r:
            return [json.loads(l) for l in r.read().decode().splitlines() if l.strip()]


# The index. State: a SQLite connection (`check_same_thread=False`, so the console's threaded server may
# query it), `state` (`"empty"`, `"catching up"`, `"live[; … unreachable][; … from mirror]"`) and
# `indexed_segments`.
class EventIndex:
    # Creates two tables: `seen (server, path)` primary key — which buckets have been ingested, keyed by the
    # real owning server and the original path; and `events (subsystem, unit, cam, epoch, t, kind, server,
    # path, fields)` with indexes on `(cam, t)`, `(subsystem, unit, t)` and `(kind, t)`. `fields` holds the
    # remaining event keys as JSON. `lost_after` is how old a resource heartbeat may be before the resource
    # counts as silent.
    def __init__(self, reader, path: str = ":memory:", wall=time.time, lost_after: float = 45.0):
        self.reader, self.wall, self.lost_after = reader, wall, lost_after
        self.db = sqlite3.connect(path, check_same_thread=False)
        self.db.executescript("""
            CREATE TABLE IF NOT EXISTS seen   (server TEXT, path TEXT, PRIMARY KEY (server, path));
            CREATE TABLE IF NOT EXISTS events (subsystem TEXT, unit TEXT, cam INTEGER, epoch INTEGER, t REAL, kind TEXT,
                                               server TEXT, path TEXT, fields TEXT);
            CREATE INDEX IF NOT EXISTS events_cam_t ON events (cam, t);
            CREATE INDEX IF NOT EXISTS events_sub_unit_t ON events (subsystem, unit, t);
            CREATE INDEX IF NOT EXISTS events_kind_t ON events (kind, t);""")
        self.state = "empty"
        self.indexed_segments = 0

    # From nothing — what a failed-over instance does first: truncate both tables, zero the counter, then
    # `tail(resources, rebuild=True)`.
    def rebuild(self, resources: dict[str, dict]) -> dict:
        """From nothing: what a failed-over instance does first."""
        self.db.executescript("DELETE FROM seen; DELETE FROM events;")
        self.indexed_segments = 0
        return self.tail(resources, rebuild=True)

    # `resources` is `resources_seen()` (`{server: heartbeat}`). Sets `state` to "catching up" on a rebuild.
    # Splits servers into live and silent by `ts`. For a silent server, try `_from_mirror`; if no live peer
    # holds copies, list it as unreachable. For a live server, walk `heartbeat["units"]` and, for each
    # bucket the resource reports, skip it if it has no events or is already in `seen`; otherwise fetch its
    # lines and insert one row each — `cam` is the event's `cam` field, or the unit id itself when the unit
    # is numeric (a numeric unit is its own `cam`; others may name one) — then mark the bucket seen, all in
    # one transaction per bucket. Any exception while talking to a live server (fresh heartbeat, server not
    # answering) marks it unreachable. Ends by setting `state` to "live" plus the unreachable and
    # from-mirror lists, and returns `{added, unreachable, from_mirror, segments}`. Because only closed
    # buckets are reported by the resource and `seen` is keyed per bucket, tailing is idempotent and each
    # bucket is read once.
    def tail(self, resources: dict[str, dict], rebuild: bool = False) -> dict:
        self.state = "catching up" if rebuild else self.state
        now = self.wall(); added = 0; unreachable = []; from_mirror = []
        live = {s: hb for s, hb in resources.items() if now - float(hb["ts"]) <= self.lost_after}
        for server, hb in sorted(resources.items()):
            if server not in live:
                if self._from_mirror(server, live):
                    from_mirror.append(server); added += self._last_mirror_added
                else:
                    unreachable.append(server)
                continue
            try:
                for sub, units in hb.get("units", {}).items():
                    for unit in units:
                        for b in self.reader.buckets(hb["url"], sub, unit):
                            if b.events == 0 or self.db.execute("SELECT 1 FROM seen WHERE server=? AND path=?", (server, b.path)).fetchone():
                                continue
                            rows = []
                            for e in self.reader.events(hb["url"], b):
                                cam = e.get("cam", int(unit) if unit.isdigit() else None)     # a numeric unit is its own `cam`; others may name one
                                rows.append((sub, unit, cam, b.epoch, float(e["t"]), e["kind"], server, b.path,
                                             json.dumps({k: v for k, v in e.items() if k not in ("t", "kind", "cam")})))
                            with self.db:
                                self.db.executemany("INSERT INTO events VALUES (?,?,?,?,?,?,?,?,?)", rows)
                                self.db.execute("INSERT INTO seen VALUES (?,?)", (server, b.path))
                            added += len(rows); self.indexed_segments += 1
            except Exception:                          # noqa: BLE001 — fresh heartbeat, server not answering
                unreachable.append(server)
        self.state = "live" + (f"; {', '.join(unreachable)} unreachable" if unreachable else "") \
                            + (f"; {', '.join(from_mirror)} from mirror" if from_mirror else "")
        return {"added": added, "unreachable": unreachable, "from_mirror": from_mirror, "segments": self.indexed_segments}

    # A silent server's closed buckets, from whichever live peer's heartbeat says `mirrors` includes it.
    # Rows are inserted under the real server and original path — only the source differed — so a later
    # `tail` from the recovered server skips them as seen. Returns False if nobody holds copies; a peer that
    # fails is skipped. Sets `_last_mirror_added` for `tail` to sum. Never used while the resource answers.
    def _from_mirror(self, server: str, live: dict[str, dict]) -> bool:
        """A silent server's closed buckets, from whichever live peer holds
        copies (its heartbeat says: mirrors). Rows are inserted under the
        REAL server: only the source differed. Never used while the resource answers."""
        self._last_mirror_added = 0
        holders = [(s, hb) for s, hb in live.items() if server in hb.get("mirrors", {})]
        if not holders:
            return False
        for peer, hb in holders:
            try:
                for b in self.reader.mirrored(hb["url"], server):
                    if b.events == 0 or self.db.execute("SELECT 1 FROM seen WHERE server=? AND path=?", (server, b.path)).fetchone():
                        continue
                    rows = []
                    for e in self.reader.mirrored_events(hb["url"], server, b):
                        cam = e.get("cam", int(b.unit) if b.unit.isdigit() else None)
                        rows.append((b.subsystem, b.unit, cam, b.epoch, float(e["t"]), e["kind"], server, b.path,
                                     json.dumps({k: v for k, v in e.items() if k not in ("t", "kind", "cam")})))
                    with self.db:
                        self.db.executemany("INSERT INTO events VALUES (?,?,?,?,?,?,?,?,?)", rows)
                        self.db.execute("INSERT INTO seen VALUES (?,?)", (server, b.path))
                    self._last_mirror_added += len(rows); self.indexed_segments += 1
            except Exception:                        # noqa: BLE001 — that peer is not answering either
                continue
        return True

    def query(self, t0: float, t1: float, cam: int | None = None, kind: str | None = None,
              subsystem: str | None = None, unit: str | None = None,
              current_epochs: dict[tuple[str, str], int] | None = None, limit: int = 1000) -> dict:
        """`current_epochs` is {(subsystem, unit): epoch} — fencing is per unit, and only the
        unit's own subsystem knows its current epoch; the index just compares."""
        sql, args = "SELECT subsystem, unit, cam, epoch, t, kind, server, path, fields FROM events WHERE t >= ? AND t < ?", [t0, t1]
        if cam is not None: sql += " AND cam = ?"; args.append(cam)
        if kind is not None: sql += " AND kind = ?"; args.append(kind)
        if subsystem is not None: sql += " AND subsystem = ?"; args.append(subsystem)
        if unit is not None: sql += " AND unit = ?"; args.append(str(unit))
        sql += " ORDER BY t LIMIT ?"; args.append(limit)
        out = []
        for sub, u, c, ep, t, k, server, path, fields in self.db.execute(sql, args):
            cur = (current_epochs or {}).get((sub, u))
            out.append({"subsystem": sub, "unit": u, "cam": c, "epoch": ep, "t": t, "kind": k, "server": server, "bucket": path,
                        "fenced": cur is not None and ep < cur, **json.loads(fields)})
        return {"events": out, "state": self.state}

    # Retention on a resource removed a segment: delete its events and its `seen` row so a re-mirrored copy
    # would not be refused. Returns the count of paths.
    def forget(self, server: str, paths: list[str]) -> int:
        """Retention on a resource removed a segment: its events go with it."""
        with self.db:
            for p in paths:
                self.db.execute("DELETE FROM events WHERE server=? AND path=?", (server, p))
                self.db.execute("DELETE FROM seen WHERE server=? AND path=?", (server, p))
        return len(paths)
